Editor/EditorCore/CutTool.cs

```diff
@@ -0,0 +1,1421 @@
+                }
+
+                case KeyCode.Escape:
+                    evt.Use();
+                    Clear();
+                    //Leave the current tool
+                    Tools.current = Tool.None;
+                    break;
+
+                case KeyCode.KeypadEnter:
+                case KeyCode.Return:
+                case KeyCode.Space:
+                    evt.Use();
+                    ExecuteCut();
+                    //Leave the current tool
+                    Tools.current = Tool.None;
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Display existing points of the cut
+        /// </summary>
+        void DoExistingPointsGUI()
+        {
+            Transform trs = m_Mesh.transform;
+            int len = m_CutPath.Count;
+
+            Event evt = Event.current;
+
+            if (evt.type == EventType.Repaint)
+            {
+                for (int index = 0; index < len; index++)
+                {
+                    Vector3 point = trs.TransformPoint(m_CutPath[index].position);
+                    float size = HandleUtility.GetHandleSize(point) * k_HandleSize;
+
+                    Handles.color = k_HandleColor;
+                    Handles.DotHandleCap(-1, point, Quaternion.identity, size, evt.type);
+                }
+
+                Handles.color = Color.white;
+            }
+        }
+
+        /// <summary>
+        /// Display current position to potentially add to the cut
+        /// </summary>
+        void DoCurrentPointsGUI()
+        {
+            Transform trs = m_Mesh.transform;
+            int len = m_CutPath.Count;
+
+            Event evt = Event.current;
+
+            if (evt.type == EventType.Repaint)
+            {
+                if (!m_CurrentPosition.Equals(Vector3.positiveInfinity))
+                {
+                    Vector3 point = trs.TransformPoint(m_CurrentPosition);
+                    if(m_SelectedIndex >= 0 && m_SelectedIndex < m_CutPath.Count)
+                        point = trs.TransformPoint(m_CutPath[m_SelectedIndex].position);
+
+                    float size = HandleUtility.GetHandleSize(point) * k_HandleSize;
+                    Handles.color = m_CurrentHandleColor;
+                    Handles.DotHandleCap(-1, point, Quaternion.identity, size, evt.type);
+                }
+                Handles.color = Color.white;
+            }
+        }
+
+        /// <summary>
+        /// Visual indications to help the user: highlighting faces, edges and vertices when snapping on them
+        /// </summary>
+        void DoVisualCues()
+        {
+            if(m_Mesh != null)
+            {
+                if(m_TargetFace == null)
+                {
+                    EditorHandleDrawing.HighlightVertices(m_Mesh, m_Mesh.sharedVertexLookup.Keys.ToArray(), false);
+                    EditorHandleDrawing.HighlightEdges(m_Mesh, m_Mesh.faces.SelectMany(f => f.edges).Distinct().ToArray(),
```

Comment:
it would be nice to avoid the `ToArray` and linq calls here if we're running this on `EventType.Repaint`